// @progress/kendo-angular-dialog/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var dialog_component_1 = require("./dialog.component");
exports.DialogComponent = dialog_component_1.DialogComponent;
var dialog_titlebar_component_1 = require("./dialog-titlebar.component");
exports.DialogTitleBarComponent = dialog_titlebar_component_1.DialogTitleBarComponent;
var dialog_actions_component_1 = require("./dialog-actions.component");
exports.DialogActionsComponent = dialog_actions_component_1.DialogActionsComponent;
var dialog_module_1 = require("./dialog.module");
exports.DialogModule = dialog_module_1.DialogModule;
var dialog_service_1 = require("./dialog.service");
exports.DialogService = dialog_service_1.DialogService;
var dialog_settings_1 = require("./dialog-settings");
exports.DialogCloseResult = dialog_settings_1.DialogCloseResult;
exports.DialogRef = dialog_settings_1.DialogRef;
exports.DialogSettings = dialog_settings_1.DialogSettings;
exports.DialogAction = dialog_settings_1.DialogAction;
